### PR TITLE
graph: color large graphs by 'none' instead of 'structure' by default (4)

### PR DESCRIPTION
**Pull request by psybuzz**
The "Color by structure" mode can be very expensive to compute,
often consistently freezing TensorBoard for minutes with some large
graphs. This introduces a new "Color by none", which does not color
groups by structure.

Opening TensorBoard's Graphs with a small graph will still default to
"structure" coloring, but when loading a large graph, it will
automatically switch to "none", requiring the user to opt-in by clicking
on the radio button.

The condition for a large graph is "10K+ nodes && 10K+ edges", but
this can be tuned. These were determined by examining several
graphs manually:

| (# nodes, # edges)  | Time computing templates (sec) | Time to load (after) | Notes
| ------------- | ------------- | -------- | ----- |
| (12, 19)  | 0.004  | <3s |
| (125, 144)  | 0.007  | <3s |
| (233, 410)  | 0.013  | <3s |
| (361, 394)  | 0.011s  | <3s | Resnet8 |
| (601, 956)  | 0.013s  | <3s | Mnist |
| (603, 955)  | 0.016s  | <3s | Recommender |
| (6K, 9K)  | 0.4s  | <3s | Cifar |
| (35K, 1,968K)  | 3.6s  | 24s |
| (160K, 660K)  | unknown (2m+)  | 1m45s |
| (224K, 229K)  | 102s  | 23s |

Screenshot
image

Diffbase: https://github.com/tensorflow/tensorboard/pull/4741
Followup: none

### CODE REVIEW

File: tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ import {PolymerElement, html} from '@polymer/polymer';
 import {customElement, observe, property} from '@polymer/decorators';
 
 import '../../../components/polymer/irons_and_papers';
+
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts:19]: Remove this extra line. It does not seem to be relevant to the internal import grouping nor VSCode's default import grouping.

File: tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts [hunk 2]
```diff
@@ -309,4 +310,29 @@ class TfGraphBoard extends LegacyElementMixin(PolymerElement) {
...
   }
+  @observe('graph')
+  _slimGraphChanged() {
+    // By default, avoid coloring by 'structure' for large graphs, since it may be very expensive.
+    // Color by 'structure' is still available to users via explicit gesture.
+    if (!this.graph) {
+      return;
+    }
+    const isGraphTooLarge =
+      Object.keys(this.graph.nodes).length > 10000 &&
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts:321]: Would you mind parameterizing this via module const or graph params? You can do it on a follow up.

File: tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts [hunk 3]
```diff
@@ -309,4 +310,29 @@ class TfGraphBoard extends LegacyElementMixin(PolymerElement) {
...
+      this.graph.edges.length > 10000;
+    if (isGraphTooLarge && this.colorBy === ColorBy.STRUCTURE) {
+      this.colorBy = ColorBy.NONE;
+      this.fire('color-by-changed', this.colorBy);
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts:325]: AFAIK, literally a setter on `this.colorBy` triggers this exact event. Was it not the case?

File: tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.ts [hunk 4]
```diff
@@ -162,6 +164,7 @@ class TfGraphDashboard extends LegacyElementMixin(PolymerElement) {
             stats="[[_stats]]"
             trace-inputs="[[_traceInputs]]"
             auto-extract-nodes="[[_autoExtractNodes]]"
+            on-color-by-changed="_onBoardColorByChanged"
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.ts:167]: Would two way binding on L149 achieve the same thing? Honest question.

**Comment by stephanwlee** [comment]
Thanks for attaching the table. It was insightful.

File: tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts (hunk 3 above)
Comment (psybuzz) [reply tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts]: Ah yes Polymer magic.  Setting `this.colorBy` only triggers this event when the property has `notify: true`.  Since `.fire()` is part of the LegacyElementMixin, I've dropped it in favor of notify: true.

File: tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.ts (hunk 4 above)
Comment (psybuzz) [reply tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.ts]: Yup, done.

I realize I forgot to update tf-graph-app, so updated that, too.

File: tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts [hunk 5]
```diff
@@ -24,6 +24,18 @@ import {LegacyElementMixin} from '../../../components/polymer/legacy_element_mix
...
+ * Some UX features, such as 'color by structure', rely on the 'template'
+ * information of a graph. This can be very expensive to compute when the
+ * graph is too large. This object's constants determine what constitutes
+ * a 'large' graph. Graphs that exceed all these constraints should not
+ * have templates computed by default.
+ */
+const MAX_GRAPH_SIZE_TO_COMPUTE_TEMPLATES = {
+  maxNodeCount: 10000,
+  maxEdgeCount: 10000,
+};
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts:37]: style thing: often, consts invert the casing :P 

i.e.,
```ts
const MaxGraphSizeToComputeTemplates = {
  MAX_NODE_COUNT: 10000,
  MAX_Edge_COUNT: 10000,
};

```

**Review by stephanwlee — APPROVED** [review]

File: tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts (hunk 5 above)
Comment (psybuzz) [reply tensorboard/plugins/graph/tf_graph_board/tf-graph-board.ts]: Done.  I'll remember for next time!

**Review by psybuzz — COMMENTED** [review]
Thanks for the reviews!
